### ISSUE TITLE
HTTP wrapper class w. global configuration

### ISSUE THREAD

**Issue by pedep** [SDFIdk/CDSETool#93]
After having spent a few minutes fixing the last pylint issues, i noticed that `query_features` does not pass the `proxies` argument to the `FeatureQuery` class. This should probably be adressed.

@subnet001 @pjonsson Having looked long at the spaghetti we are weaving trying to pass arguments from the api function to the util functions that actually need them, i think we should consider separating some of this logic into a purpose-built class.

I was thinking something along the lines of this:

```python

# http.py
class HTTP
  proxies = None

  def get(...):
    requests.get(..., proxies = HTTP.proxies)
    
# main.py

from cdsetool.http import HTTP

HTTP.proxies = {
  # proxy conf
}

query_features("Sentinel2") # uses HTTP as backend instead of requests
```

This way we pass the configs directly into where they are needed via the static variable, instead of having to pass along variables through half a dozen methods. Only downside is no per-request proxies.

_Originally posted by @pedep in https://github.com/SDFIdk/CDSETool/issues/92#issuecomment-1983307927_

**Comment by pjonsson** [comment]
I had a look at the code and there are a few things I don't understand:
1. When `Credentials` is constructed, `__exchange_credentials()` is called and that performs HTTP requests. Since that happens first in `download_features`, those requests are made without a HTTP proxy, which is likely to go straight into the company firewall and fail. So `Credentials` constructor should take a proxies parameter as well? If it did, `get_session()` could configure the proxy before returning the session, and the proxy code from download.py could be removed.
1. In query.py, `_query_url()` has a default value for proxies. Since it is an internal function, and the only in-tree caller already has a default value for its own proxies parameter, nothing will ever use the default value of `_query_url`, and same comment for `_get_describe_doc`.

I can't see why someone would need to have different proxies for each request, my impression is that if you are stuck behind a proxy, you need it for all requests.

I just started learning Python so I'm not sure I understand the details of the suggestion in this issue, but `HTTP` is never passed to `query_features`, so it must be some kind of global variable? Personally, I'm not stuck behind a proxy, so the solution for proxies can be arbitrarily bad without making me unhappy, but my experience is that with time, more things will end up inside the HTTP object since it is already there, and all that state will make things harder to test, which in turn will become a maintenance issue.

I also see the need for a simple API for casual users, so maybe the current `query_features` should be renamed to `_query_features` and take a million different parameters, and a new `query_features` is introduced that deals with accessing the global state and then calling `_query_features` with the right parameters?

**Comment by pedep** [comment]
> When Credentials is constructed, __exchange_credentials() is called and that performs HTTP requests. Since that happens first in download_features, those requests are made without a HTTP proxy, which is likely to go straight into the company firewall and fail. So Credentials constructor should take a proxies parameter as well? If it did, get_session() could configure the proxy before returning the session, and the proxy code from download.py could be removed.

Yes, `Credentials` should take the proxy parameters in the constructor as well for this to work. Using credentials to construct a session with proxies pre-configured in might not be a bad idea.

> In query.py, _query_url() has a default value for proxies. Since it is an internal function, and the only in-tree caller already has a default value for its own proxies parameter, nothing will ever use the default value of _query_url, and same comment for _get_describe_doc.

While you are correct, i think its better just to have the defaults. They do not impact performance, but ensure that anyone relying on the internal API will not be affected by the change.

> I can't see why someone would need to have different proxies for each request, my impression is that if you are stuck behind a proxy, you need it for all requests.

I'm of the same opinion. I think we should go for a single proxy configuration, and re-evaluate when enough users request otherwise.
The `requests` module has a way to set the proxy config globally, but i think it would be better to be able to set it library-wide for `cdsetool`. The users might also be calling internal company API's that are behind the firewall, which would get rejected if going through the proxy.

> I just started learning Python so I'm not sure I understand the details of the suggestion in this issue, but HTTP is never passed to query_features, so it must be some kind of global variable?

My intention was to redirect every http call through the `HTTP` class. This would require every call to `requests` to be replaced with an equivalent call to our `HTTP` class instead. 

> but my experience is that with time, more things will end up inside the HTTP object since it is already there, and all that state will make things harder to test, which in turn will become a maintenance issue.

We should definitely be careful when expanding the `HTTP` module. My intention was to use the `HTTP` class as a static class, with only static methods and variables. Hopefully the state will stay simple this way, as it should be configured at application start and left alone while running.

> I also see the need for a simple API for casual users, so maybe the current query_features should be renamed to _query_features and take a million different parameters, and a new query_features is introduced that deals with accessing the global state and then calling _query_features with the right parameters?

I was working on a simplified API for users in #31 some time ago, but stopped because i was going vacationing. I should get back to that branch and finish it up.

**Comment by pjonsson** [comment]
> Yes, `Credentials` should take the proxy parameters in the constructor as well for this to work. Using credentials to construct a session with proxies pre-configured in might not be a bad idea.

It sounds like you can foresee this part, can you put together that PR? I think having that in the tree would simplify the scope for this ticket.
 
> > but my experience is that with time, more things will end up inside the HTTP object since it is already there, and all that state will make things harder to test, which in turn will become a maintenance issue.
> 
> We should definitely be careful when expanding the `HTTP` module. My intention was to use the `HTTP` class as a static class, with only static methods and variables. Hopefully the state will stay simple this way, as it should be configured at application start and left alone while running.

I was aware that our past experiences with these kind of scenarios were probably different, so that was why I suggested that the current public functions are made internal and get the HTTP object as one of its parameters, and new public functions are introduced that just passes its arguments and the global HTTP object along to the internal functions.

As an additional benefit, that would keep the internal functions testable without having to play tricks with modifying the global HTTP object, and the headache of trying to undo the modifications to the global HTTP object upon test failure to avoid cascading test failures.

> I was working on a simplified API for users in #31 some time ago, but stopped because i was going vacationing. I should get back to that branch and finish it up.

My comment was just an off the cuff guess with "It's probably good for casual users", I personally wouldn't use the API for casual users and have no indications someone else is waiting for such an API.

Even with a simplified API, I still think solid core functionality is essential for the new user experience, and off the top of my head, the hanging downloads, incomplete zip files, token expiration exceptions, and my most recent `ChunkedEncodingError: ('Connection broken: IncompleteRead(15965393 bytes read, 36747223 more expected)', IncompleteRead(15965393 bytes read, 36747223 more expected))` seem like things that would scare casual users.
